pkg/asset/tls/certkey.go

```diff
@@ -0,0 +1,153 @@
+package tls
+
+import (
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"fmt"
+	"net"
+	"path/filepath"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/types"
+)
+
+// CertKey contains the private key and the cert that's
+// signed by the parent CA.
+type CertKey struct {
+	// assetStock is a reference to the store so that
+	// we can get the InstallConfig instance.
+	assetStock *asset.Stock
+
+	// Common fields.
+	Subject      pkix.Name
+	Validity     time.Duration
+	KeyUsages    x509.KeyUsage
+	ExtKeyUsages []x509.ExtKeyUsage
+	KeyFileName  string
+	CertFileName string
```

Comment:
I'd rather use separate assets for the key and the cert.  I guess most folks who want to bring their own keys will also bring their own certs, but having separate assets would support us cert'ing folks who bring only their own keys.  A potentially more-useful reason for splitting certs from keys is that we could recycle keys and only rebuild certs if something feeding the cert changed (e.g. the DNS names because the user decided to change that config and rebuild their assets).

Reply:
In Tectonic we were going to target just allowing a user to show up with a CA (or assets to generate intermediate CA), but that was it (they wouldn't get to provide component level certificates). 

I don't know if that's acceptable from the Openshift product side. IMO letting user provide component level certs is just asking for them to mess it up (and rotation may not work as expected). But I've heard various things across the spectrum in various meetings, from:

- We generate everything (offline root + in-cluster CA). No option to customize
- We generate everything, but a user can provide is a CSR endpoint that we can use to ask for an in-cluster CA (and we'd plug endpoint into installer).
- User can provide cluster-CA or root CA used to generating cluster-CA (a-la Tectonic plans).
- User can provide all component certs.

@smarterclayton / @eparis -- any specific constraints here we should be working from? My position is going to be the less options we provide (and the more we control) the better - but not always realistic.